### PR TITLE
WT-10031 Create a CLI for the timestamp simulator

### PR DESCRIPTION
**Pull request by Siddhartha8899**

### CODE REVIEW

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 1]
```diff
@@ -27,13 +27,152 @@
...
+            case 5:
+                interface_commit_transaction();
+                break;
+            case 6:
+                interface_rollback_transaction();
+                break;
+            case 7:
+                print_rules();
+                break;
+            case 8:
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: You may want a `default` case here too.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 2]
```diff
@@ -27,13 +27,152 @@
...
+                exit = true;
+            }
+        } catch (std::string &exception_str) {
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: Can this string be `const`?

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 3]
```diff
@@ -158,21 +104,82 @@ interface_rollback_transaction()
...
-print_rules()
+int
+main(int argc, char *argv[])
 {
+    bool exit = false;
+
+    do {
+        std::cout << std::endl;
+
+        std::cout << "1: Session Management " << std::endl;
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:129]: You can make a vector of choices and then a loop that would print the index + 1 and the content of the vector.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 4]
```diff
@@ -158,21 +104,82 @@ interface_rollback_transaction()
...
+        std::cout << "4: [Session] begin_transaction()" << std::endl;
+        std::cout << "5: [Session] commit_transaction()" << std::endl;
+        std::cout << "6: [Session] prepare_transaction()" << std::endl;
+        std::cout << "7: [Session] rollback_transaction()" << std::endl;
+        std::cout << "8: [Session] timestamp_transaction()" << std::endl;
+        std::cout << "9: [Session] query_timestamp()" << std::endl;
+        std::cout << "10: print rules for timestamps" << std::endl;
+        std::cout << "11: exit" << std::endl;
+
+        int choice = choose_num(1, 11, "timestamp_simulator >>");
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:141]: 11 would become the size of the vector.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 5]
```diff
@@ -158,21 +104,82 @@ interface_rollback_transaction()
...
+            case 8:
+                interface_timestamp_transaction();
+                break;
+            case 9:
+                interface_session_query_timestamp();
+                break;
+            case 10:
+                print_rules();
+                break;
+            case 11:
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:175]: You may want a `default` statement here.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                exit = true;
+            }
+        } catch (std::string &exception_str) {
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:178]: Can the string be `const`?

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        std::cout << "2: [Conn] set_timestamp()" << std::endl;
+        std::cout << "3: [Conn] query_timestamp()" << std::endl;
+        std::cout << "4: [Session] begin_transaction()" << std::endl;
+        std::cout << "5: [Session] commit_transaction()" << std::endl;
+        std::cout << "6: [Session] prepare_transaction()" << std::endl;
+        std::cout << "7: [Session] rollback_transaction()" << std::endl;
+        std::cout << "8: [Session] timestamp_transaction()" << std::endl;
+        std::cout << "9: [Session] query_timestamp()" << std::endl;
+        std::cout << "10: print rules for timestamps" << std::endl;
+        std::cout << "11: exit" << std::endl;
```
Comment (Sean04) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:139]: I think a reset or restart option would be useful.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 3 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: Updated in [d3b07ca](https://github.com/wiredtiger/wiredtiger/pull/8417/commits/d3b07cad2974c45eb19a02631df03808cff9758d)

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 4 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: Updated in [d3b07ca](https://github.com/wiredtiger/wiredtiger/pull/8417/commits/d3b07cad2974c45eb19a02631df03808cff9758d)

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 5 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:185]: As discussed we check for this to ensure that the range is always between the values in the case statement. Hence, let's not worry about this.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 6 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: Updated in [d3b07ca](https://github.com/wiredtiger/wiredtiger/pull/8417/commits/d3b07cad2974c45eb19a02631df03808cff9758d)

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 7 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: Good idea, I will probably get to this at the end of the PR.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 5 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:251]: Updated in [199f738](https://github.com/wiredtiger/wiredtiger/pull/8417/commits/199f738890b1943ec1e0325002fbb7bdd5ecc0ba)

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 8]
```diff
@@ -76,19 +77,116 @@ choose_num(int min, int max, const std::string &cli_str)
...
+            }
+
+            std::string hex_ts;
+            bool ts_supported;
+            int ret = conn->query_timestamp(config, hex_ts, ts_supported);
+
+            if (hex_ts == "0")
+                hex_ts = "not specified";
+
+            if (ret != 0)
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:239]: Should we check this before checking the output `hex_ts`?

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 9]
```diff
@@ -285,8 +285,44 @@ interface_commit_transaction(session_simulator *session)
...
+    std::vector<std::string> options;
+    options.push_back("Prepare timestamp");
+    options.push_back("Run operation");
+    options.push_back("<- go back");
+
+    do {
+        try {
+            print_options(options);
+
+            int choice = choose_num(1, options.size(), "Commit transaction >>");
```
Comment (etienneptl) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:359]: Did you try to enter other than numbers when this function is called?

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 7 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: I think a reset option can be added later if this tool is used often. Rn, it does not seem worth the effort. 
If someone wants to reset, they can exit and start over.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 8 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:236]: Thanks, updated in [f78f929](https://github.com/wiredtiger/wiredtiger/pull/8417/commits/f78f929a04e0d3cc79069fffd5a18b52f823476f)

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 9 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:359]: Yup, letters, empty string. It works fine.

Comment (Sean04) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:359]: It returns "Choose a number between 1 and 3" here if you enter some other character.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp [hunk 10]
```diff
@@ -0,0 +1,645 @@
...
+                config += "roundup_timestamps=(read=true)";
+                break;
+            case 3:
+                config += "roundup_timestamps=(prepare=true)";
+                break;
+            case 4: {
+                int ret = session->begin_transaction(config);
+
+                if (ret != 0)
+                    throw "begin_transaction failed with return value: " + std::to_string(ret);
```
Comment (Sean04) [inline test/simulator/timestamp/simulator_interface/simulator_interface.cpp:283]: I think we should go back to the main menu in this case.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 9 above)
Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:359]: In commit_transaction the range is 1 to 4. Hence, it says `Choose a number between 1 and 4`.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 10 above)
Comment (Sean04) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:283]: You could just return where you catch the error.

Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:283]: Agreed! Just pushed a commit to do the same. [30d0e9d](https://github.com/wiredtiger/wiredtiger/pull/8417/commits/30d0e9d60cc7c0a517d4c38444c64374f54e269a)

Comment (Siddhartha8899) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:283]: Even if it runs successfully better is to go to the main menu, cause there's nothing else you could do.

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 7 above)
Comment (Sean04) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp]: Yeah that's fine

File: test/simulator/timestamp/simulator_interface/simulator_interface.cpp (hunk 10 above)
Comment (Sean04) [reply test/simulator/timestamp/simulator_interface/simulator_interface.cpp:283]: That too!

**Review by etienneptl — APPROVED** [review]

**Review by Sean04 — APPROVED** [review]
LGTM

**Pull request merged**